tests/asgi/test_asgi_servers.py

```diff
@@ -0,0 +1,185 @@
+            pass
+
+    def test_post_read_bounded_stream(self, server_base_url):
+        body = testing.rand_string(_SIZE_1_KB / 2, _SIZE_1_KB)
+        resp = requests.post(server_base_url + 'bucket', data=body, timeout=_REQUEST_TIMEOUT)
+        assert resp.status_code == 200
+        assert resp.text == body
+
+    def test_post_read_bounded_stream_no_body(self, server_base_url):
+        resp = requests.post(server_base_url + 'bucket', timeout=_REQUEST_TIMEOUT)
+        assert not resp.text
+
+    def test_sse(self, server_base_url):
+        resp = requests.get(server_base_url + 'events', timeout=_REQUEST_TIMEOUT)
+        assert resp.status_code == 200
+
+        events = resp.text.split('\n\n')
+        assert len(events) > 2
+        for e in events[:-1]:
+            assert e == 'data: hello world'
+
+        assert not events[-1]
+
+
+@contextmanager
+def _run_server_isolated(process_factory, host, port):
+    # NOTE(kgriffs): We have to use subprocess because uvicorn has a tendency
+    #   to corrupt our asyncio state and cause intermittent hangs in the test
+    #   suite.
+    print('\n[Starting server process...]')
+    server = process_factory(host, port)
+
+    time.sleep(0.2)
+    startup_succeeded = (server.poll() is None)
+    print('\n[Server process start {}]'.format('succeeded' if startup_succeeded else 'failed'))
+
+    if startup_succeeded:
+        yield server
+
+    print('\n[Sending SIGTERM to server process...]')
+    server.terminate()
+
+    try:
+        server.communicate(timeout=10)
+    except subprocess.TimeoutExpired:
+        server.kill()
+        server.communicate()
+
+    assert server.returncode == 0
+    assert startup_succeeded
+
+
+def _uvicorn_factory(host, port):
+    # NOTE(vytas): uvicorn+uvloop is not (well) supported on PyPy at the time
+    #   of writing.
+    loop_options = ('--http', 'h11', '--loop', 'asyncio') if _PYPY else ()
+    options = (
+        '--host', host,
+        '--port', str(port),
+
+        '_asgi_test_app:application'
+    )
+
+    return subprocess.Popen(
+        ('uvicorn',) + loop_options + options,
+        cwd=_MODULE_DIR,
+    )
+
+
+def _daphne_factory(host, port):
+    return subprocess.Popen(
+        (
+            'daphne',
+
+            '--bind', host,
+            '--port', str(port),
+
+            '--verbosity', '2',
+            '--access-log', '-',
+
+            '_asgi_test_app:application'
+        ),
+        cwd=_MODULE_DIR,
+    )
+
+
+@pytest.fixture(params=[_uvicorn_factory, _daphne_factory])
+def server_base_url(request):
+    process_factory = request.param
+
+    # NOTE(kgriffs): This facilitates parallel test execution as well as
+    #   mitigating the problem of trying to reuse a port that the system
+    #   hasn't cleaned up yet.
+    # NOTE(kgriffs): Use our own Random instance because we don't want
+    #   pytest messing with the seed.
+    server_port = _random.randint(50000, 60000)
```

Comment:
We'll probably need to make this more robust in the future, because it may randomly clash with other stuff on the system; although the probability is low, it may clash between tests as well.